mrc-2511 Emit changes in runWorkflowRun to persist then in workflow metadata

EmmaLRussell opened the pull request:
This branch brings `runWorkflowRun`, the finalise workflow page, into line with `runWorkflowReport` by having it emit any workflow metadata changes to the wizard rather than managing them internally. This means the that metadata will be updated ready to be submitted when user presses 'Run workflow'. 

It also means that it should be possible to go back to the previous step, then forward again to the finalise change, and see that any changes previously made have been retained. This required updating the changelog and instances components to support mounting with initial values.

EmmaLRussell commented on src/app/static/src/js/components/runWorkflow/runWorkflowRun.vue:
```diff
@@ -133,24 +144,23 @@
-                if(this.workflowName) {
-                    this.$emit("update", {name: this.workflowName})
-                }
+                this.$emit("update", {name: workflowName});
```
We still want to emit the change even if it is not valid -we'll get odd behvaviour if the wizard's metadata isn't updated to match the input - it will just invalid and unsubmittable.

mwoodbri commented on src/app/static/src/js/components/runReport/instances.vue:
```diff
@@ -64,13 +69,14 @@
             selectInitialInstance: function () {
                 if (this.instances && this.showInstances) {
```
Not new, but `this.instances` seems logically to be the same as `this.showInstances`

mwoodbri commented on src/app/static/src/js/components/runReport/instances.vue:
```diff
@@ -64,13 +69,14 @@
+                            this.$set(this.selectedInstances, key, initialInstances[key] || instances[key][0]);
                         }
                     }
+                    this.$emit("selectedValues", this.selectedInstances);
```
Just to note that the logic here differs slightly from `changeLog.vue`, where we don't emit an event if initial value is used to populate selected value. I can see why we do here - because `selectedValues` could in theory contain a combination of initial and non-initial values. So I guess you could change the `changeLog` code to emit an event, even if it's unlikely that users of the component will be expecting or need such an event.

EmmaLRussell replied:
Yes, it is a little inconsistent. Maybe it would be better just to emit here if any new selected values were actually chosen during initialisation?

mwoodbri reviewed (changes requested):
One change necessary: if I type in a workflow name (thus enabling submit button), and then click Back and Next, the name is retained but the button becomes disabled. This might pre-date this PR but obviously an issue.

Tangentially related, but we don't seem to have a Selenium test that touches the final page of wizard - would be good to extend what we currently have, especially in anticipation of enabling the submit button itself.

Two other inline comments that I'll leave to your discretion.

EmmaLRussell commented:
- fixed the issue with the submit button being disabled on reload, by emitting an initial valid event
- added a few checks into the Selenium test for the finalise step
- removed the double check for `showInstances` in the instances component
- made the instances component consistent with changelog - it only emits an event on mount if it initialised at least instance value which was not already provided

mwoodbri commented on src/app/static/src/tests/components/runWorkflow/runWorkflowRun.test.ts:
```diff
@@ -323,6 +362,7 @@ describe(`runWorkflowRun`, () => {
 
             expect((wrapper.find("#changelogMessage").element as HTMLTextAreaElement).value).toBe("Workflow message");
             expect((wrapper.find("#changelogType").element as HTMLSelectElement).value).toBe("public");
+            done();
```
Good catch...

mwoodbri reviewed (approved):
Great stuff
